src/braket/reservations/reservations.py

```diff
@@ -0,0 +1,70 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from collections.abc import Callable
+from contextlib import contextmanager
+
+from braket.aws.aws_device import AwsDevice, AwsDeviceType
+from braket.devices.device import Device
+from braket.devices.local_simulator import LocalSimulator
+
+
+@contextmanager
+def Reservation(device: Device, reservation_arn: str | None = None) -> Device | None:
+    """
+    A context manager that temporarily modifies the `run` method of a `device` object
+    to include a specified `reservation_arn` in all calls within the context. This is useful
+    for ensuring that all quantum task run within the context are associated with a given
+    reservation ARN (Amazon Resource Name).
+
+    Args:
+        device (Device): The Braket device for which you have a reservation ARN.
+        reservation_arn (str | None): The Braket Direct reservation ARN to be applied to all
```

Comment:
What would be the use case of this with no reservation arn?

Reply:
So that a customer doesn't have to modify their code if they don't have an ARN. Example

```
def run_tasks(device, circuit, shots, reservation_arn=None):
    with Reservation(device, reservation_arn=reservation_arn) as reservation:
        task1 = device.run(circuit, shots)
        task2 = device.run(circuit, shots)
        return task1, task2
```
vs 

```
def run_tasks(device, circuit, shots, reservation_arn=None):
    if reservation_arn is None:
        task1 = device.run(circuit, shots)
        task2 = device.run(circuit, shots)
        return task1, task2
    elif reservation_arn:
        with Reservation(device, reservation_arn=reservation_arn) as reservation:
            task1 = device.run(circuit, shots)
            task2 = device.run(circuit, shots)
        return task1, task2
```

Reply:
updated to make reservation_arn required now, but its still allowed to be None.